src/triage/component/postmodeling/contrast/model_evaluator.py

```diff
@@ -0,0 +1,1155 @@
+            f'''
+            SELECT model_id,
+                   as_of_date,
+                   metric,
+                   feature_column,
+                   value,
+                   threshold_unit,
+                   threshold_value
+            FROM test_results.crosstabs_test
+            WHERE model_id = {self.model_id}
+            ''', con=conn)
+
+        return model_crosstabs
+
+    def preds_matrix(self, path):
+        '''
+        Load predictions matrices using the catwalk.storage.ProjectStorage.
+        This class allow the user to define a project path that can be either a
+        system local path or an s3 path and it will handle itself this
+        different infrastructures.
+
+        Once defined, we can pass this object to the MatrixStorageEngine that
+        will read this object and return a MatrixSotre object with a set of
+        handy methods to handle matrices
+
+        Arguments:
+            path: project path to initiate the ProjectStorage object
+        '''
+        cache = self.__dict__.setdefault('_preds_matrix_cache', {})
+        try:
+            return cache[path]
+        except KeyError:
+            pass
+
+        storage = ProjectStorage(path)
+        matrix_storage = MatrixStorageEngine(storage).get_store(self.pred_matrix_uuid)
+        mat = matrix_storage.matrix
+
+        # Merge with predictions table and return complete matrix
+        merged_df = pd.merge(mat,
+                             self.predictions,
+                             on=['entity_id', 'as_of_date'],
+                             how='inner',
+                             suffixes=('test', 'pred'))
+
+        cache[path] = merged_df
+        return merged_df
+
+    def train_matrix(self, path):
+        '''
+        Load predictions matrices using the catwalk.storage.ProjectStorage.
+        This class allow the user to define a project path that can be either a
+        system local path or an s3 path and it will handle itself this
+        different infrastructures.
+
+        Once defined, we can pass this object to the MatrixStorageEngine that
+        will read this object and return a MatrixSotre object with a set of
+        handy methods to handle matrices
+
+        Arguments:
+            path: project path to initiate the ProjectStorage object
+        '''
+        cache = self.__dict__.setdefault('_preds_matrix_cache', {})
+        try:
+            return cache[path]
+        except KeyError:
+            pass
+
+        storage = ProjectStorage(path)
+        matrix_storage = MatrixStorageEngine(storage).get_store(self.train_matrix_uuid)
+        mat = matrix_storage.matrix
+
+        # Merge with predictions table and return complete matrix
+        merged_df = mat.merge(self.predictions,
+                             on='entity_id',
+                             how='inner',
+                             suffixes=('test', 'pred'))
+
+        cache[path] = mat
+        return mat
```

Comment:
This method, `train_matrix`, looks almost identical to `preds_matrix`, **and** *steals its cache*. What's going on?